### PR TITLE
Add slash commands support

### PR DESCRIPTION
**Pull request by BanTheNons**
This PR adds support for Discord's slash commands and Button interactions

### CODE REVIEW

File: src/commands/Admin/channel.ts [hunk 1]
```diff
@@ -29,7 +32,7 @@ const command: Command = {
             value: "all"
         }]
     }],
-    execute(interaction: Discord.CommandInteraction) {
+    async execute(interaction: Discord.CommandInteraction) {
```
Comment (ImRodry) [inline src/commands/Admin/channel.ts:35]: Same as before

File: src/commands/Admin/eval.ts [hunk 2]
```diff
@@ -1,9 +1,3 @@
-const { loadingColor, errorColor, successColor, neutralColor, blurple, listeningStatuses, watchingStatuses, playingStatuses } = require("../../config.json")
-const fetch = require("node-fetch")
-const { flag, code, name, countries } = require('country-emoji')
-const fs = require("fs")
-const country = require("countryjs")
-import { db as mongoDb } from "../../lib/dbclient"
```
Comment (ImRodry) [inline src/commands/Admin/eval.ts:6]: These importa are here for a reason, don’t remove them

File: src/commands/Admin/eval.ts [hunk 3]
```diff
@@ -15,20 +9,15 @@ const command: Command = {
...
   }],
-  async execute(interaction: Discord.CommandInteraction, getString: (path: string, variables?: { [key: string]: string | number } | string, cmd?: string, lang?: string) => any) {
-    const me = interaction.member,
-      guild = interaction.guild,
-      channel = interaction.channel as Discord.TextChannel,
-      db = mongoDb,
-      discord = Discord,
-      client = Client
-
+  async execute(interaction: Discord.CommandInteraction) {
```
Comment (ImRodry) [inline src/commands/Admin/eval.ts:20]: Same as above

File: src/index.ts [hunk 4]
```diff
@@ -14,9 +14,12 @@ export interface Command extends Discord.ApplicationCommandData {
     cooldown?: number,
     allowDM?: true,
     allowTip?: false,
+    roleWhitelist?: string[],
+    roleBlacklist?: string[],
+    channelWhitelist?: string[],
     dev?: true,
     category?: string,
-    execute(interaction: Discord.CommandInteraction, getString?: (path: string, variables?: { [key: string]: string | number } | string, cmd?: string, lang?: string) => any): any
+    execute(interaction: Discord.CommandInteraction, lang?: string): Promise<any>
```
Comment (ImRodry) [inline src/index.ts:22]: Lang? Also not all executes return promises

File: src/listeners/ready.ts [hunk 5]
```diff
@@ -1,11 +1,69 @@
-import { client } from "../index.js"
+import { client, Command } from "../index.js"
 import stats from "../events/stats.js"
 import inactives from "../events/inactives.js"
 import { listeningStatuses, watchingStatuses, playingStatuses } from "../config.json"
+import { ApplicationCommand, ApplicationCommandPermissionData, Collection } from "discord.js"
```
Comment (ImRodry) [inline src/listeners/ready.ts:5]: Just import Discord as a whole, we do that across the entire bot

File: src/listeners/ready.ts [hunk 6]
```diff
@@ -1,11 +1,69 @@
...
 
 client.once("ready", async () => {
     console.log("Ready!")
 
+    const publishCommand = async (command: Command) => {
+        if (command.allowDM) {
+            //Create a global command
+            var cmd = await client.application?.commands.create({
```
Comment (ImRodry) [inline src/listeners/ready.ts:13]: Don’t use `var`
Also we can’t create the command every time, only the first time this update is published

File: src/listeners/ready.ts [hunk 7]
```diff
@@ -1,11 +1,69 @@
...
+                name: command.name,
+                description: command.description,
+                defaultPermission: command.defaultPermission,
+                options: command.options
+            })
+        } else {
+            //Create a guild wide command
+            var cmd = await client.guilds.cache.get('841233609012543489')?.commands.create({
```
Comment (ImRodry) [inline src/listeners/ready.ts:21]: Wrong ID?

File: src/listeners/ready.ts [hunk 8]
```diff
@@ -1,11 +1,69 @@
...
+                name: command.name,
+                description: command.description,
+                defaultPermission: command.defaultPermission,
+                options: command.options
+            }) 
+        }
+
+        let permissions: ApplicationCommandPermissionData[] = []
+        if (command.roleWhitelist) { //Add whitelisted roles
+            command.roleWhitelist.forEach(id => {
```
Comment (ImRodry) [inline src/listeners/ready.ts:31]: You can avoid the if by using command.roleWhitelist?.forEach

File: src/listeners/ready.ts [hunk 9]
```diff
@@ -1,11 +1,69 @@
...
+                permissions.push({
+                    type: 1,
+                    id,
+                    permission: true
+                })
+            })
+        } else if (command.roleBlacklist) { //Add blacklisted roles
+            command.roleBlacklist.forEach(id => {
```
Comment (ImRodry) [inline src/listeners/ready.ts:39]: Same as above

File: src/listeners/ready.ts [hunk 10]
```diff
@@ -1,11 +1,69 @@
...
+            })
+        }
+
+        if (permissions) {
+            await cmd?.setPermissions(permissions)
+        }
+    }
+
+    //Fetch slash commands
+    const commands: Collection<string, ApplicationCommand> | undefined = await client.application?.commands?.fetch()
```
Comment (ImRodry) [inline src/listeners/ready.ts:54]: No need to set the type, this is done automatically

File: src/listeners/ready.ts [hunk 11]
```diff
@@ -1,11 +1,69 @@
...
+    if (!commands) {
+        client.commands.forEach(async (command: Command) => {
+            await publishCommand(command)
+        })
+    } else {
+        client.commands.forEach(async (command: Command) => {
+            if (!commands.some(cmd => cmd.name === command.name)) { //Chech if the command is published
```
Comment (ImRodry) [inline src/listeners/ready.ts:61]: What if the command is edited?

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: src/commands/Admin/eval.ts (hunk 3 above)
Comment (ImRodry) [reply src/commands/Admin/eval.ts:26]: Totally not what I’m talking about. Don’t change the constants and don’t remove unused declarations on eval, they’re there for a reason. And yes the function does not need to be async

Comment (BanTheNons) [reply src/commands/Admin/eval.ts:26]: Oops, I didn't think about the fact that they would be used inside eval.

Comment (ImRodry) [reply src/commands/Admin/eval.ts:26]: Still not fixed

Comment (BanTheNons) [reply src/commands/Admin/eval.ts]: They should be fixed now.

File: src/commands/Admin/bulksend.ts [hunk 12]
```diff
@@ -18,7 +20,7 @@ const command: Command = {
     description: "The amount of messages to send in bulk",
     required: true
   }],
-  execute(interaction: Discord.CommandInteraction) {
+  async execute(interaction: Discord.CommandInteraction) {
```
[the remark this answers is not in the document]
Comment (BanTheNons) [reply src/commands/Admin/bulksend.ts]: Reverted

File: src/listeners/ready.ts [hunk 13]
```diff
@@ -1,11 +1,66 @@
...
+                description: command.description,
+                defaultPermission: command.defaultPermission,
+                options: command.options
+            }) 
+        }
+
+        let permissions: Discord.ApplicationCommandPermissionData[] = []
+        command.roleWhitelist?.forEach(id => { //Add whitelisted roles
+            permissions.push({
+                type: 1,
```
Comment (ImRodry) [inline src/listeners/ready.ts:32]: Don't use numeric types, instead use STRING or ROLE

File: src/listeners/ready.ts [hunk 14]
```diff
@@ -1,11 +1,66 @@
...
+                id,
+                permission: true
+            })
+        })
+        command.roleBlacklist?.forEach(id => { //Add blacklisted roles
+            permissions.push({
+                type: 1,
```
Comment (ImRodry) [inline src/listeners/ready.ts:39]: same as above

File: src/listeners/ready.ts [hunk 15]
```diff
@@ -1,11 +1,66 @@
...
+            var cmd = await client.guilds.cache.get('841233609012543489')?.commands.create({
+                name: command.name,
+                description: command.description,
+                defaultPermission: command.defaultPermission,
+                options: command.options
+            }) 
+        }
+
+        let permissions: Discord.ApplicationCommandPermissionData[] = []
+        command.roleWhitelist?.forEach(id => { //Add whitelisted roles
```
Comment (ImRodry) [inline src/listeners/ready.ts:30]: if there is a whitelist, the verified role should not have permission. Along with that, the everyone role should not have permission for any command except for verify

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: src/listeners/ready.ts (hunk 15 above)
Comment (BanTheNons) [reply src/listeners/ready.ts:30]: I'll make it so it makes `defaultPermission` false if `roleWhitelist` exists.

Comment (BanTheNons) [reply src/listeners/ready.ts:30]: Alright, I wrote this:
```js
command.defaultPermission ? true : command.defaultPermission !== undefined ? false : command.roleWhitelist ? false : true
```
It looks absolutely horrible, so if you have a way of making it cleaner please do

Comment (ImRodry) [reply src/listeners/ready.ts:30]: Can you explain what exactly you’re trying to do? I believe you could use `??` although I’m not 100% sure

Comment (BanTheNons) [reply src/listeners/ready.ts:30]: Yep, that works

File: src/listeners/ready.ts (hunk 13 above)
Comment (BanTheNons) [reply src/listeners/ready.ts]: Fixed

File: src/listeners/ready.ts (hunk 14 above)
Comment (BanTheNons) [reply src/listeners/ready.ts]: Fixed

**Comment by ImRodry** [comment]
This PR was not merged due to the author's inactivity

**Pull request closed without merging**